substrate/latch

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef SUBSTRATE_LATCH
+#define SUBSTRATE_LATCH
+
+#include <atomic>
+#include <condition_variable>
+#include <cstddef>
+#include <mutex>
+
+namespace substrate
+{
+	struct latch_t final
+	{
+	private:
+		mutable std::mutex latchMutex{};
+		mutable std::condition_variable waitZero{};
+		std::atomic<std::size_t> count{};
+
+	public:
+		explicit latch_t(std::size_t expected) : count{expected} { }
+		~latch_t() = default;
+
+		void countDown() noexcept { if (count) --count; }
```

Comment:
Appreciated you were keeping this formatted like in pcat, but for substrate, we'd split this onto multiple lines as in:
```suggestion
		void countDown() noexcept
		{
			if (count)
				--count;
		}
```